release notes for the 1.15 release

joemoe opened the pull request.

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+and ensuring a smooth transition for both Flink users and for external connector
+developers:
+
+1. all internal Flink connectors that depend on `flink-connector-base` now shade
```
this was actually partially reverted; cc @AHeise @afedulov

afedulov replied:
Yes, the commit with the associated release note was reverted.

afedulov replied:
Some background: unfortunately some classes within the `flink-connector-base`,  such as `DeliveryGuarantee`, that are public and can be used in the user code and need to be first moved into `flink-core`. The same applies to the `HybridSource`.

gaoyunhaii replied:
Thanks @afedulov for the information! I'll then first remove this issue from the release note.

zentol reviewed (commented):
Why is this missing FLINK-20845?

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+There are three changes in Flink 1.15 that require updating dependency names when 
+upgrading from earlier versions:
+
+* The newly introduced module `flink-table-planner-loader` (FLINK-25128) could 
+  replace the legacy `flink-table-planner_2.12`. As a consequence, `flink-table-uber` 
+  has been split into `flink-table-api-java-uber`, `flink-table-planner(-loader), 
+  and `table-runtime`. Besides, the artifactId of `flink-sql-client` has no Scala 
+  suffix (_2.11 / _2.12) anymore, and Scala users need to explicitly add a 
+  dependency to `flink-table-api-scala` or `flink-table-api-scala-bridge`.
```
This isn't phrased in a user-friendly way. We have to come up with something along the lines of "if you used A, now use B".
IMO we also shouldn't combine so many things into 1 item (split/suffix removal/scala opt-in).

gaoyunhaii replied:
Thanks @zentol for the review! I'll split them into different items and give a summary table for the relationship.

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+This ticket resolves an issue that during state migration between Flink versions
+the wrong serializer might have been picked.
+
+When upgrading from Flink 1.13.x please immediately choose 1.14.3 or higher and 
+skip 1.14.0, 1.14.1, 1.14.2 because all are affected and it might prevent your 
+job from starting.
```
Should be able to remove this given that we're talking about 1.15.

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+`TableEnvironment.fromTableSource`, `TableEnvironment.sqlUpdate`, and 
+`TableEnvironment.explain` have been removed. Please use the provided 
+alternatives introduced in 
+[FLIP-84](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=134745878).
```
IMO we shouldn't link to FLIPs in user-facing content. If we want to refer them to anything, then I guess it should be the javadocs that should contain migration instructions.

joemoe replied:
changed the text, will also add some javadocs links.

zentol commented on docs/content/release-notes/flink-1.15.md:
````diff
@@ -0,0 +1,379 @@
+EXPLAIN STATEMENT SET BEGIN ... END;
+```
+
+### [Check & possible fix decimal precision and scale for all Aggregate functions (FLINK-24809)](https://issues.apache.org/jira/browse/FLINK-functions)
````
broken jira link

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+implement custom formats for `FileSystemTableSource` might need to verify the 
+implementation and make sure to implement `ProjectableDecodingFormat` if necessary.
+
+### [Update user document for batch window tvf (FLINK-24760)](https://issues.apache.org/jira/browse/FLINK-24760)
+
+All window table-valued functions can now also be used in batch mode.
+
+### [Add user document for Window Deduplication (FLINK-24656)](https://issues.apache.org/jira/browse/FLINK-24656)
+
+Flink SQL now also supports Window Deduplication as a new window table-valued function.
+`WindowDeduplication` is a deduplication operator which removes rows that duplicate
+over a set of columns, keeping only the first one or the last one for each window
+and other partitioned keys.
```
I don't think these 2 belong in here. They are ultimately new features, and not a cause for issues during upgrades.

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+
+## Connectors
+
+### [(Introducing GCS FileSystem Support and) Update GCS documentation (FLINK-25577)](https://issues.apache.org/jira/browse/FLINK-25577)
```
also not relevant for upgrades imo

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+
+### [Merge flink-connector-testing into flink-connector-test-utils (FLINK-25712)](https://issues.apache.org/jira/browse/FLINK-25712)
+
+The `flink-connector-testing` module has been removed and should use
+`flink-connector-test-utils` module instead.
```
```suggestion
The `flink-connector-testing` module has been removed and users should use
`flink-connector-test-utils` module instead.
```

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+The `flink-connector-testing` module has been removed and should use
+`flink-connector-test-utils` module instead.
+
+### [Support to disable caching missing key for lookup cache (FLINK-24861)](https://issues.apache.org/jira/browse/FLINK-24861)
```
not relevant for these release notes

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+configured and disabled. More details can be found 
+[in the documentation](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/deployment/config/#retryable-cleanup).
+
+### [Introduce common interfaces for cleaning up local and global job data (FLINK-25432)](https://issues.apache.org/jira/browse/FLINK-25432)
+
+Utilizes the newly added component `JobResultStore` (see 
+[FLINK-25431](https://issues.apache.org/jira/browse/FLINK-25431)) to pick up
+already finished but not properly cleaned up jobs after a failover.
```
This is an internal thing that doesn't belong into the release notes afaik. @XComp

XComp replied:
You're actually right. The important things are the retry configuration (FLINK-25433) and the file-based JRS (FLINK-25431). I will remove the release notes from FLINK-25432.

XComp replied:
Done... ✅

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+Reduce down-scaling delay in reactive mode through signaling between `TaskManager` 
+and `JobManager`.
+
+### [FLIP-187: Adaptive Batch Scheduler (FLINK-24892)](https://issues.apache.org/jira/browse/FLINK-24892)
+
+In 1.15, we introduced a new scheduler to Flink: Adaptive Batch Scheduler. The 
+new scheduler can automatically decide parallelisms of job vertices for batch
+jobs, according to the size of data volume each vertex needs to process.
+
+More information about this feature and how to enable it could be found in 
+[the documentation](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/elastic_scaling/#adaptive-batch-scheduler.)
```
new feature; should be reserved to blog post

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+TM-per-job metrics are now released upon last slot release rather than task, i.e. 
+may stay on TM longer.
+
+### [Issue in exceptions API(/jobs/:jobid/exceptions) in flink 1.13.2 (FLINK-25096)](https://issues.apache.org/jira/browse/FLINK-25096)
```
@XComp please change the jira title to be more descriptive

XComp replied:
Good point - I updated the Jira issue title. Sorry for not thinking about it while losing the issue.

XComp replied:
Done ✅

XComp commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,379 @@
+### [Implement file-based JobResultStore (FLINK-25431)](https://issues.apache.org/jira/browse/FLINK-25431)
+
+Introduces a file-based version of the `JobResultStore` with new configuration
+parameters `job-result-store.storage-path` and `job-result-store.delete-on-commit`. 
```
```suggestion
parameters [job-result-store.storage-path](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/deployment/config/#job-result-store-storage-path) and [job-result-store.delete-on-commit](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/deployment/config/#job-result-store-delete-on-commit). 
```

joemoe commented:
thanks for all the good catches @zentol. i addressed most of your comments. @gaoyunhaii will also give a pass and then i will also add the remaining links.

gaoyunhaii commented:
Hi @zentol for the [FLINK-20845](https://issues.apache.org/jira/browse/FLINK-20845) it seems the current curl way in https://cwiki.apache.org/confluence/display/FLINK/Creating+a+Flink+Release |  Prepare for the release seems indeed not include this issue: https://issues.apache.org/jira//rest/api/2/search?jql=project%20%3D%20FLINK%20AND%20%22Release%20Note%22%20is%20not%20EMPTY%20and%20fixVersion%20%3D%201.15.0

We'll have a double check with the web UI for the missing issues.

gaoyunhaii commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,614 @@
+    flink-table-api-java-bridge
+    flink-test-utils
+    flink-yarn
+    flink-table-uber
```
Also cc @zentol @twalthr for a double check for this part~

zentol commented:
> We'll have a double check with the web UI for the missing issues.

Based on the json response jira only returns the first 50 items. you can add `&maxResults=100` to the URL to get all items.

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,614 @@
+    flink-connector-elasticsearch5
+    flink-connector-elasticsearch6
+    flink-connector-elasticsearch7
+    flink-connector-gcp-pubsub
+    flink-connector-hbase-1.4
+    flink-connector-hbase-2.2
+    flink-connector-hbase-base
+    flink-connector-jdbc
+    flink-connector-kafka
+    flink-connector-kinesis
+    flink-connector-nifi
+    flink-connector-pulsar
+    flink-connector-rabbitmq
+    flink-connector-testing
+    flink-connector-twitter
+    flink-connector-wikiedits
+    flink-container
+    flink-dstl-dfs
+    flink-gelly
+    flink-hadoop-bulk
+    flink-kubernetes
+    flink-runtime-web
+    flink-sql-connector-elasticsearch6
+    flink-sql-connector-elasticsearch7
+    flink-sql-connector-hbase-1.4
+    flink-sql-connector-hbase-2.2
+    flink-sql-connector-kafka
+    flink-sql-connector-kinesis
+    flink-sql-connector-rabbitmq
+    flink-state-processor-api
+    flink-statebackend-rocksdb
+    flink-streaming-java
+    flink-table-api-java-bridge
+    flink-test-utils
+    flink-yarn
+    flink-table-uber
+    flink-table-runtime
+    flink-orc
+    flink-orc-nohive
+    flink-parquet
```
```suggestion
    flink-cep
    flink-clients
    flink-connector-elasticsearch-base
    flink-connector-elasticsearch6
    flink-connector-elasticsearch7
    flink-connector-gcp-pubsub
    flink-connector-hbase-1.4
    flink-connector-hbase-2.2
    flink-connector-hbase-base
    flink-connector-jdbc
    flink-connector-kafka
    flink-connector-kinesis
    flink-connector-nifi
    flink-connector-pulsar
    flink-connector-rabbitmq
    flink-container
    flink-dstl-dfs
    flink-gelly
    flink-hadoop-bulk
    flink-kubernetes
    flink-runtime-web
    flink-sql-connector-elasticsearch6
    flink-sql-connector-elasticsearch7
    flink-sql-connector-hbase-1.4
    flink-sql-connector-hbase-2.2
    flink-sql-connector-kafka
    flink-sql-connector-kinesis
    flink-sql-connector-rabbitmq
    flink-state-processor-api
    flink-statebackend-rocksdb
    flink-streaming-java
    flink-table-api-java-bridge
    flink-test-utils
    flink-yarn
    flink-table-uber
    flink-table-runtime
    flink-orc
    flink-orc-nohive
    flink-parquet
```
These don't exist in 1.15 anymore.

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,614 @@
+    flink-cep
+    flink-clients
+    flink-connector-elasticsearch-base
+    flink-connector-elasticsearch5
```
```suggestion
```
Not part of 1.15.

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,614 @@
+    flink-connector-nifi
+    flink-connector-pulsar
+    flink-connector-rabbitmq
+    flink-connector-testing
+    flink-connector-twitter
+    flink-connector-wikiedits
```
```suggestion
```
Not part of 1.15.

zentol replied:
```suggestion
```
Not part of 1.15.

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,614 @@
+  the APIs that you are using. This should primarily affect users of the Scala `DataStream/CEP` APIs.
+
+* A lot of modules have lost their Scala suffix. 
+  Any dependency on one of the modules listed below needs to be updated. 
```
```suggestion
```

gaoyunhaii commented:
Very thanks @zentol for the double reviewing, I updated the PR accordingly. 

For the `curl` command, I'll also update the release document~

zentol commented on docs/content/release-notes/flink-1.15.md:
````diff
@@ -0,0 +1,597 @@
+    flink-orc-nohive
+    flink-parquet
+    ```
+* For Table / SQL users, the new module `flink-table-planner-loader` replaces `flink-table-planner_2.12`
+  and avoids the need for a Scala suffix. As a consequence, `flink-table-uber` has been split into `flink-table-api-java-uber`,
+  `flink-table-planner(-loader)`, and `flink-table-runtime`. Scala users need to explicitly add a dependency
+  to `flink-table-api-scala` or `flink-table-api-scala-bridge`. For backwards compatibility, users can still
+  swap it with `flink-table-planner_2.12` located in `opt/`.
````
```suggestion
* For Table / SQL users, the new module `flink-table-planner-loader` replaces `flink-table-planner_2.12`
  and avoids the need for a Scala suffix. For backwards compatibility, users can still
  swap it with `flink-table-planner_2.12` located in `opt/`.
  `flink-table-uber` has been split into `flink-table-api-java-uber`,
  `flink-table-planner(-loader)`, and `flink-table-runtime`. Scala users need to explicitly add a dependency
  to `flink-table-api-scala` or `flink-table-api-scala-bridge`.
```

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,597 @@
+application. This depends on the data types that you have been using in your
+application.
+
+Also, the Scala Shell/REPL has been removed in
+[FLINK-24360](https://issues.apache.org/jira/browse/FLINK-24360).
```
```suggestion
The Scala Shell/REPL has been removed in
[FLINK-24360](https://issues.apache.org/jira/browse/FLINK-24360).
```

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,597 @@
+
+##### [FLINK-24858](https://issues.apache.org/jira/browse/FLINK-24858)
+
+This ticket resolves an issue that during state migration between Flink versions
+the wrong serializer might have been picked.
```
Is this relevant for the upgrade? Sounds more like a plain bug fix that users don't need to be aware of.

gaoyunhaii replied:
Yes, it should be a plain bug fix. I'll then remove it.

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,597 @@
+
+## Table API & SQL
+
+#### Make the legacy behavior disabled by default 
```
```suggestion
#### Disable the legacy casting behavior by default 
```

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,597 @@
+The legacy casting behavior has been disabled by default. This might have 
+implications on corner cases (string parsing, numeric overflows, to string 
+representation, varchar/binary precisions). Set 
+`table.exec.legacy-cast-behaviour=ENABLED` to restore the old behavior.
```
Are there plans to remove the legacy-cast-behavior (i.e., is it deprecated)?

gaoyunhaii replied:
Checking with the SQL guys the legacy behavior should not be deprecated in short period since the difference might be very large. It might be removed in, for example, 2.0

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,597 @@
+
+##### [FLINK-24809](https://issues.apache.org/jira/browse/FLINK-24809)
+
+This changes the result of a decimal `SUM()` with retraction and `AVG()` between
+1.14.0 and 1.14.1. It restores the behavior of 1.13 to be consistent with Hive/Spark.
```
So...if you're upgrading from 1.13.X/1.14.1+ nothing changes?

gaoyunhaii replied:
The issue itself is merged to 1.14 / 1.13, thus I think for high minor version of 1.13.x it should be such case. I'll remove the part irrelevant to 1.15.

gaoyunhaii replied:
By checking with the SQL guys, the story would be like 
- If we mark 1.13 as the base
- In 1.14, we do some modification, makes it partly compatible with Hive/spark and partly compatible with sqlserver
- In 1.15, we revert the part that compatible with sqlserver back to the behavior of 1.13, makes it compatible with Hive/Spark as a whole

I'll try modifying the release notes.

zentol commented on docs/content/release-notes/flink-1.15.md:
````diff
@@ -0,0 +1,597 @@
+CONVERT_TZ
+```
+
+#### Support IS JSON for Table API 
+
+##### [FLINK-16501](https://issues.apache.org/jira/browse/FLINK-16501)
+
+`IS JSON` does not return `NULL` anymore but always `FALSE` (even if the argument is `NULL`).
````
the ticket title and description don't really fit. The title makes it seem sound like a new feature, while the description seems more like a behavior change.

zentol commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,597 @@
+
+##### [FLINK-24864](https://issues.apache.org/jira/browse/FLINK-24864)
+
+TM-per-job metrics are now released upon last slot release rather than task, i.e. 
+may stay on TM longer.
```
This was meant to be added as a suggestion :/

gaoyunhaii replied:
I'll modify this part accordingly~

gaoyunhaii commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,597 @@
+##### [FLINK-24323](https://issues.apache.org/jira/browse/FLINK-24323)
+
+`ElasticsearchXSinkBuilder` supersedes `ElasticsearchSink.Builder` and provides at-least-once writing with the
+new unified sink interface supporting both batch and streaming mode of DataStream API.To upgrade, please stop with savepoint.
```
I also think we could remove this statement.

gaoyunhaii commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,597 @@
+
+##### [FLINK-24864](https://issues.apache.org/jira/browse/FLINK-24864)
+
+TM-per-job metrics are now released upon last slot release rather than task, i.e. 
+may stay on TM longer.
```
```suggestion
Job metrics on the TaskManager are now removed when the last slot is released, rather than the last task.
This means they may be reported for a longer time than before and when no tasks are running on the TaskManager.
```

gaoyunhaii commented on docs/content/release-notes/flink-1.15.md:
```diff
@@ -0,0 +1,597 @@
+prevents Flink from finalizing the job cleanup. The retry functionality can be
+configured and disabled. More details can be found 
+[in the documentation](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/deployment/config/#retryable-cleanup).
+
+#### Introduce common interfaces for cleaning up local and global job data 
+
+##### [FLINK-25432](https://issues.apache.org/jira/browse/FLINK-25432)
+
+Utilizes the newly added component `JobResultStore` (see 
+[FLINK-25431](https://issues.apache.org/jira/browse/FLINK-25431)) to pick up
+already finished but not properly cleaned up jobs after a failover.
```
```suggestion
```

gaoyunhaii commented:
Very thanks @zentol for the review! I have updated the PR according to the comments~

gaoyunhaii commented:
Very thanks @zentol for the review! will merge~

gaoyunhaii commented:
Merged on master via 1.15
